ntripping/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+        let auth = self.auth.as_ref().map(|auth| {
+            let auth = base64::encode(format!("{}:{}", auth.username, auth.password));
+            (header::AUTHORIZATION, format!("Basic {auth}"))
+        });
+        let gga = self
+            .ntrip_gga
+            .map(|gga| (header::NTRIP_GGA, gga.to_string(false)));
+
+        std::iter::once((
+            header::SWIFT_CLIENT_ID,
+            self.client_id
+                .as_deref()
+                .unwrap_or("00000000-0000-0000-0000-000000000000")
+                .to_owned(),
+        ))
+        .chain(auth)
+        .chain(gga)
+    }
+}
+
+pub struct Auth {
+    username: String,
+    password: String,
+}
+
+impl Auth {
+    pub fn new(username: impl Into<String>, password: impl Into<String>) -> Self {
+        Self {
+            username: username.into(),
+            password: password.into(),
+        }
+    }
+}
+
+impl fmt::Debug for Auth {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Auth")
+            .field("username", &self.username)
+            .field("password", &"***")
+            .finish()
+    }
+}
+
+#[derive(Debug)]
+pub struct Connection {
+    send: SendHalf,
+    recv: ReceiveHalf,
+}
+
+impl Connection {
+    pub fn split(self) -> (SendHalf, ReceiveHalf) {
+        (self.send, self.recv)
+    }
+}
+
+impl Stream for Connection {
+    type Item = Result<Bytes, ReceiveError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Pin::new(&mut self.recv).poll_next(cx)
+    }
+}
+
+impl<T> Sink<T> for Connection
+where
+    T: Into<Bytes>,
+{
+    type Error = SendError;
+
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        <SendHalf as Sink<T>>::poll_ready(Pin::new(&mut self.send), cx)
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
+        <SendHalf as Sink<T>>::start_send(Pin::new(&mut self.send), item)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        <SendHalf as Sink<T>>::poll_flush(Pin::new(&mut self.send), cx)
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        <SendHalf as Sink<T>>::poll_close(Pin::new(&mut self.send), cx)
+    }
+}
+
+#[derive(Debug)]
+pub struct ReceiveHalf {
+    body: body::Incoming,
+}
+
+impl Stream for ReceiveHalf {
+    type Item = Result<Bytes, ReceiveError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match Pin::new(&mut self.body).poll_frame(cx) {
+            Poll::Ready(Some(Ok(frame))) => {
+                let chunk = frame.into_data();
+                if let Some(chunk) = &chunk {
+                    tracing::trace!(length = chunk.len(), "recv");
```

Comment:
Are these `impl` blocks here primarily to introduce logging?

Reply:
This one is because `body::Incoming` doesn't implement `Stream` (I think because of trailer response headers, but we don't have to deal with those in ntrip afaik). The one on `SendHalf` is just forwarding to the `Sink` impl on `mpsc::Sender`. That is just to hide the channel from the public API, could also return the `mpsc::Sender<Bytes>` directly, or an `impl Sink` instead of a named `SendHalf`. Opted against `mpsc::Sender` because it's an implementation detail, and I like a named type because we can add extra methods to it, vs needing an extension trait like with `sbp::iter_messages`